be/src/types/bitmap_value.cpp

```diff
@@ -923,91 +923,164 @@ void BitmapValue::_convert_to_smaller_type() {
+        if ((offset > 0 && offset >= cardinality) || (offset < 0 && std::abs(offset) > cardinality)) {
+            return 0;
+        }
+        int64_t abs_offset = offset;
+        if (offset < 0) {
+            abs_offset = cardinality + offset;
+        }
 
-    int64_t count = 0;
-    int64_t offset_count = 0;
-    auto it = _bitmap->begin();
-    for (; it != _bitmap->end() && offset_count < abs_offset; ++it) {
-        ++offset_count;
+        std::vector values(_set->begin(), _set->end());
+        std::sort(values.begin(), values.end());
+
+        int64_t count = 0;
+        for (auto idx = abs_offset; idx < values.size() && count < len; ++idx, ++count) {
+            ret_bitmap->add(values[idx]);
+        }
+        return count;
     }
-    for (; it != _bitmap->end() && count < len; ++it, ++count) {
-        ret_bitmap->add(*it);
+    default:
+        DCHECK_EQ(_type, BITMAP);
+        size_t cardinality = _bitmap->cardinality();
+        if ((offset > 0 && offset >= cardinality) || (offset < 0 && std::abs(offset) > cardinality)) {
+            return 0;
+        }
+        int64_t abs_offset = offset;
+        if (offset < 0) {
+            abs_offset = cardinality + offset;
+        }
+
+        int64_t count = 0;
+        int64_t offset_count = 0;
+        auto it = _bitmap->begin();
+        for (; it != _bitmap->end() && offset_count < abs_offset; ++it) {
+            ++offset_count;
+        }
+        for (; it != _bitmap->end() && count < len; ++it, ++count) {
+            ret_bitmap->add(*it);
+        }
+        return count;
     }
-    return count;
 }
 
 int64_t BitmapValue::bitmap_subset_limit_internal(const int64_t& range_start, const int64_t& limit,
-
                                                   BitmapValue* ret_bitmap) {
-    bool is_reverse = false;
-
-    int64_t abs_limit = limit;
-    if (limit < 0) {
-        is_reverse = true;
-        abs_limit = -limit;
+    switch (_type) {
+    case EMPTY:
+        return 0;
+    case SINGLE: {
+        if (_sv < range_start || limit == 0) {
+            return 0;
+        } else {
+            ret_bitmap->add(_sv);
+            return 1;
+        }
     }
-
-    int64_t count = 0;
-
-    if (is_reverse) {
-        detail::Roaring64Map::const_iterator start = _bitmap->begin();
-        detail::Roaring64Map::const_iterator end = _bitmap->begin();
-
-        int64_t offset = 0;
-        for (; end != _bitmap->end() && offset < abs_limit && *end <= range_start;) {
-            ++end;
-            ++offset;
+    case SET: {
+        std::vector values(_set->begin(), _set->end());
+        std::sort(values.begin(), values.end());
+
+        int64_t count = 0;
+        if (limit < 0) {
+            int64_t abs_limit = -limit;
+            auto it = std::lower_bound(values.begin(), values.end(), range_start);
+            for (; it != values.begin() && count < abs_limit; --it, ++count) {
+                ret_bitmap->add(*it);
+            }
```

Comment:
end() is invalid, but begin() is valid, so I think there may be bugs

Reply:
Is is possible that `it` is `values.end()`? If `range_start` is greater than all the values in `values`, then `it` will be `values.end()`.